### ISSUE TITLE
Monitoring of Concourse

### ISSUE THREAD

**Issue by keithkroeger** [concourse/concourse#2128]
# Feature Request

## Running out of Containers and Volume Limits

We see an issue in BOSH deployments where the number of containers used by BOSH reaches or exceeds the limit of 250.  While the monit processes for Concourse watch for the fact that various elements of the concourse application (TSA, ATC, worker,...) are running, the monitoring does not seem to watch for limits or other health statistics.

## A Modest Proposal

It might be helpful if, for example, the monit script noted that a worker is unhealthy if the number of containers has reached some limit.

What we've had to do is create separate operations pipelines that monitor for these cases and either react or alert.

**Comment by jama-pivotal** [comment]
@keithkroeger +1 to this issue. The Concourse team also has the [Oxygen Mask](https://github.com/concourse/oxygen-mask) project for using a separate concourse instance to monitor service objectives. Unfortunately the project is probably not generically usable and distributable at the moment, but it might give you some ideas down the road

**Comment by cirocosta** [comment]
Hey,

This sounds like something we're aiming to solve with https://github.com/concourse/concourse/issues/2753 (although that issue is k8s specific, it can totally fit the BOSH lifecycle).

*tl;dr: we started with adding a healthcheck endpoint to the worker (that can be used by monit as well) that just pings garden & baggageclaim, but then next, we can do some more interesting checks - like, trying to create a minimal container & volume (which in the case you described, would make the probe fail).*

That way, workers could provide to their corresponding runtimes (monit / kubelet) the necessary information about their readiness to serve their workloads.

wdyt?

thx!

**Comment by aledeganopix4d** [comment]
It is now possible to obtain the number of running containers per worker through the metric: `concourse_workers_containers` exported by the ATC.
That metric allows to cover all the use case described by this feature request. So I believe this can be closed.

**Comment by jamesmealing** [comment]
Is this metric available in the DB anywhere @aledegano? Or is it collated dynamically from the heartbeat or something else?

**Comment by taylorsilva** [comment]
@jamesmealing it's available through the prometheus metric emitter. You need to configure the emitter https://concourse-ci.org/metrics.html

The metrics are kept in-memory and emitted on an interval. Some stats may be based on data in the db

**Comment by jamesmealing** [comment]
Understood. Thank you @taylorsilva. 🙂
